// fis-parser-marked-template/lib/config.js
module.exports = {
    // highlightjs: http://highlightjs.readthedocs.org/en/latest/api.html#configure-options
    hljs: {
        tabReplace: '    '
    },
    // marked: https://github.com/chjj/marked#markedmarkdownstring-options-callback
    marked: {
        gfm: true,        
        breaks: true
    },
    /*
    file: some.md
    <!--CONFIG
    {
        "author": "nimo",
        "template": "index.html"
    }
    CONFIG-->
    template = "./" + "index.html"
    */
    templateDir: './',
    // if not find "./default.html" , use this .
    defaultTemplate: function () {/*!
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="renderer" content="webkit">
    <meta http-equiv="X-UA-Compatible" content="IE=edge" >
    <title><%- title %></title>
</head>
<body>
<%- content %>
</body>
</html>
    */},
    codeCompile: {
        // coffee: function (code) {
        //     var coffee = require('coffee-script')
        //     return {
        //         type: 'js',
        //         code: coffee.compile(code)
        //     }
        // }
    }
}
